packages/next/next-server/lib/router/router.ts

```diff
@@ -757,7 +757,7 @@ export default class Router implements BaseRouter {
     // Default to scroll reset behavior unless explicitly specified to be
     // `false`! This makes the behavior between using `Router#push` and a
     // `<Link />` consistent.
-    options.scroll = !!(options.scroll ?? true)
+    options.scroll = !!(options.scroll ?? (options.shallow ? false : true))
```

Comment:
Disable `scroll` for `shallow` links by default. Related to this change: https://github.com/vercel/next.js/pull/21437